doc/neps/nep-0034.rst

```diff
@@ -0,0 +1,87 @@
+==========================================
+NEP 34 — Disallow default ``dtype=object``
+==========================================
+
+:Author: Matti Picus
+:Status: Draft
+:Type: Standards Track
+:Created: 2019-10-10
+
+
+Abstract
+--------
+
+``np.array([<value])`` with no ``dtype`` keyword argument will sometimes
+default to an ``object``-dtype array. Change the behaviour to raise a
+`ValueError` instead.
+
+Motivation and Scope
+--------------------
+
+Users who specify lists-of-lists when creating a `numpy.ndarray` via
+``np.array`` may mistakenly pass in lists of different lengths. Currently we
+accept this input and create a ragged array with ``dtype=object``. This can be
+confusing, since it is rarely what is desired. Changing the default dtype
+detection to never return ``object`` will force users who actually wish to
+create ``object`` arrays to specify that explicitly, see for instance `issue
+5303`_.
+
+Detailed description
+--------------------
+
+After this change, ragged array creation must explicitly define a dtype:
+
+    a = np.array([1, 2], [1])
+    ValueError: cannot guess the desired dtype from the input
+
+    a = np.array([1, 2], [1], dtype=object)
```

Comment:
I actually don't think we even want to allow this. I'd consider `np.ragged_object_array(list_of_lists, depth=n)` or similar, to force a specific depth

Reply:
adopted